python/ray/autoscaler/_private/aws/config.py

```diff
@@ -719,6 +723,51 @@ def _get_key(key_name, config):
         raise exc
 
 
+def _configure_from_network_interfaces(config):
+    for cfg_key in NODE_KIND_CONFIG_KEYS.values():
+        config = _configure_node_type_from_network_interface(config, cfg_key)
+    return config
+
+
+def _configure_node_type_from_network_interface(config, node_type):
+    node_cfg = config[node_type]
+    if "NetworkInterfaces" not in node_cfg:
+        return config
+    _configure_subnets_and_groups_from_network_interfaces(node_cfg)
+    return config
+
+
+def _configure_subnets_and_groups_from_network_interfaces(node_cfg):
+    # If NetworkInterfaces are defined, SubnetId and SecurityGroupIds
+    # can't be specified in head/worker node config.
+    conflict_keys = ["SubnetId", "SubnetIds", "SecurityGroupIds"]
+    if any(conflict in node_cfg for conflict in conflict_keys):
```

Comment:
I think we'll want to do this validation against the combined `available_node_types` + `worker_nodes` config?

Reply:
I think the updated code at https://github.com/amzn/amazon-ray/blob/229a8d390d04c673624cc85c3c7babb267290293/python/ray/autoscaler/_private/aws/config.py#L840-L917 should play nicely w/ `available_node_types` on Ray 1.4.0, but I haven't had a chance to run integration tests yet to confirm. I'll update this PR with these changes once testing confirms that they work correctly, but it would be good to have you take a quick look as well.

Reply:
@pdames  Missed this comment!
Sorry for prematurely closing the PR! Have reopened.

It would be great if this were updated for compatibility with 1.4.0.
cc @ckw017 who made the 1.4.0 available_node_types updates to this PR

Along with the appropriate changes to aws/config.py we should probably in this PR revert/modify the temporary change in AWSNodeProvider.create_node which forces use of NetworkInterfaces https://github.com/ray-project/ray/pull/16816
cc @xcharleslin 

(Using NetworkInterfaces by default is fine, but main thing is to make sure create_node is consistent aws/config.py!)

Thanks!!

Reply:
Thanks for the update - I'll get this merged on top of the latest from main to resolve discrepancies between it and https://github.com/ray-project/ray/pull/16816.

Reply:
Merged. In this case, I reverted #16816 in favor of users now having the ability to be very explicit about whether or not they want to associate public IP addresses with a given network interface. However, if we'd like to keep the behavior from #16816 as the default when network interfaces aren't specified, then that is also a pretty straightforward change.

Reply:
I think the decision made in this PR makes sense cc @xcharleslin